sentinel/src/transitions/accept_patient_reports.js

```diff
@@ -114,6 +114,45 @@ const addRegistrationToDoc = (doc, registrations) => {
   }
 };
 
+const findValidRegistration = (doc, registrations) => {
+    const visitReportedDate = doc.reported_date;
+
+    for (var i = 0; i < registrations.length; i++) {
+      var registration = registrations[i];
+      if (registration.scheduled_tasks) {
+        for (var j = 0; j < registration.scheduled_tasks.length; j++) {
+          var task = registration.scheduled_tasks[j];
+          if (['delivered', 'sent'].includes(task.state)) {
+            var nextTask = registration.scheduled_tasks[j + 1] || { due: moment(visitReportedDate).add(1, 'd') };
+            if (nextTask && moment(nextTask.due) > moment(visitReportedDate)) {
+              // We loop through tasks. Once we find one that has either the status "delivered" or "sent" with
+              // a future task with a due date that is after the visit reported date then we have found the task 
+              // linked to the visit. We always link if this is the last scheduled task delivered or sent.
+              registration.scheduled_tasks[j].report_uuid = doc._id;
+              return registration;
+            }
+          }          
+        }
+      }
+    } 
+
+    return null;
+};
+
+const addReportUUIDToRegistration = (doc, registrations, callback) => {
+  if (registrations.length) {
+    const validRegistration = findValidRegistration(doc, registrations);
+
+    if (validRegistration) {
+      db.medic.put(validRegistration, callback);
+    } else {
+      callback();
```

Comment:
You don't need both `callback()` calls if you use `return` on your `db.medic.put`:

```js
const addReportUUIDToRegistration = (doc, registrations, callback) => {
  if (registrations.length) {
    const validRegistration = findValidRegistration(doc, registrations);
    if (validRegistration) {
      return db.medic.put(validRegistration, callback);
    }
  }

  callback();
}
```

In fact, due to how truthy and falsy work in JavaScript, you don't even need two if blocks:

```js
const addReportUUIDToRegistration = (doc, registrations, callback) => {
  // Or have findValidRegistrations guard against registration.length being undefined or 0
  const validRegistration = registrations.length && findValidRegistration(doc, registrations);
  if (validRegistration) {
    return db.medic.put(validRegistration, callback);
  }

  callback();
}
```